contracts/exchange/contracts/src/MixinMatchOrders.sol

```diff
@@ -166,47 +331,126 @@ contract MixinMatchOrders is
             rightTakerAssetAmountRemaining
         );
 
-        // Calculate fill results for maker and taker assets: at least one order will be fully filled.
-        // The maximum amount the left maker can buy is `leftTakerAssetAmountRemaining`
-        // The maximum amount the right maker can sell is `rightMakerAssetAmountRemaining`
-        // We have two distinct cases for calculating the fill results:
-        // Case 1.
-        //   If the left maker can buy more than the right maker can sell, then only the right order is fully filled.
-        //   If the left maker can buy exactly what the right maker can sell, then both orders are fully filled.
-        // Case 2.
-        //   If the left maker cannot buy more than the right maker can sell, then only the left order is fully filled.
-        if (leftTakerAssetAmountRemaining >= rightMakerAssetAmountRemaining) {
-            // Case 1: Right order is fully filled
-            matchedFillResults.right.makerAssetFilledAmount = rightMakerAssetAmountRemaining;
-            matchedFillResults.right.takerAssetFilledAmount = rightTakerAssetAmountRemaining;
-            matchedFillResults.left.takerAssetFilledAmount = matchedFillResults.right.makerAssetFilledAmount;
-            // Round down to ensure the maker's exchange rate does not exceed the price specified by the order.
-            // We favor the maker when the exchange rate must be rounded.
-            matchedFillResults.left.makerAssetFilledAmount = _safeGetPartialAmountFloor(
-                leftOrder.makerAssetAmount,
-                leftOrder.takerAssetAmount,
-                matchedFillResults.left.takerAssetFilledAmount
-            );
+        // Maximally fill the orders and pay out profits to the matcher in one or both of the maker assets.
+        if (withMaximalFill) {
```

Comment:
Nit - this `if` statement is pretty lengthy. It could help with readability to break up it up into two helper functions.